kge/indexing.py

```diff
@@ -2,6 +2,112 @@
 from collections import defaultdict, OrderedDict
 import numba
 import numpy as np
+from typing import Dict, List, Iterator, Tuple
+
+
+class KvsAllIndexDict:
+    def __init__(self, triples: torch.Tensor, key_cols: List, value_column: int,
+                 default_factory: type):
+        """
+        Construct KvsAllIndexDict
+        Args:
+            triples: data
+            key_cols: column indicators of data, giving the keys of the dictionary
+            value_column: column indicator of data, giving the values of the dictionary
+            default_factory: default return type
+        """
+        self.key_cols = key_cols
+        self.value_column = value_column
+        self.data_sorted = self.sort_data_by_keys(triples, key_cols, value_column)
+        self.index, self.offset = self.construct_index()
```

Comment:
offsets & why needed?

Reply:
Offsets used in KvsAll Training in call to index.index_tensors() (used to be index_KvsAll_to_torch) . Stored here so that we don't have to change too much in the training job.